### PR TITLE
Refactor Matrix -> Slack subsitutions

### LINKED ISSUE #183: Substrings of pills should not turn into Slack mentions

### PR DESCRIPTION
**Pull request by Half-Shot**
Hopefully fixes #183

### CODE REVIEW

File: src/Main.ts [hunk 1]
```diff
@@ -552,33 +556,35 @@ export class Main {
...
                     try {
-                        await room.onMatrixEdit(ev);
+                        success = await room.onMatrixEdit(ev);
                     } catch (e) {
                         log.error("Failed processing matrix edit: ", e);
                         endTimer({outcome: "fail"});
                         return;
                     }
-                    endTimer({outcome: "success"});
                     return;
```
Comment (auscompgeek) [inline src/Main.ts]: Don't you still want to `endTimer` here?

File: src/tests/unit/substitutionsTest.ts [hunk 2]
```diff
@@ -26,29 +29,219 @@ const displayMap = {
...
+                    body: "@room Hello everyone!",
+                },
+                sender: "@alice:localhost",
+            }, fakeMain, "");
+            expect(res).to.deep.equal({
+                link_names: true,
+                text: "@channel Hello everyone!",
+                username: "@alice:localhost",
+            });
+        });
```
Comment (auscompgeek) [inline src/tests/unit/substitutionsTest.ts]: This looks like the same test twice?

Comment (Half-Shot) [reply src/tests/unit/substitutionsTest.ts]: Copy paste bug, good catch.

File: src/substitutions.ts [hunk 3]
```diff
@@ -341,11 +326,16 @@ export function replacementFromDisplayMap(str: string, displaymap: IDisplayMap):
...
  */
-async function plainTextSlackMentions(main: Main, s: string, roomId: string) {
-    return replacementFromDisplayMap(
-        s,
-        await substitutions.getDisplayMap(main, roomId),
-    );
+async function plainTextSlackMentions(main: Main, body: string, teamId: string) {
+    const users = await main.datastore.getAllUsersForTeam(teamId);
+    for (const user of users) {
+        const displayName = `@${user.display_name}`;
```
Comment (Cadair) [inline src/substitutions.ts]: So you went with `@` prefix?

File: src/substitutions.ts [hunk 4]
```diff
@@ -34,37 +39,22 @@ interface IDisplayMap {
...
 }
 
-interface IFirstWordMap {
-    [firstword: string]: [IDisplayMap];
+interface PillItem {
+    id: string;
+    text: string;
 }
 
 export interface ISlackToMatrixResult {
```
Comment (Cadair) [inline src/substitutions.ts]: This interface feels like it's named backwards? It's only used in `matrixToSlack`?

File: src/tests/unit/substitutionsTest.ts [hunk 5]
```diff
@@ -26,29 +29,219 @@ const displayMap = {
...
+                },
+                sender: "@alice:localhost",
+            }, fakeMain, "");
+            expect(res).to.deep.equal({
+                link_names: true,
+                text: "Let's talk about &lt;html&gt; and the &amp; character",
+                username: "@alice:localhost",
+            });
+        });
+        it ("should replace @channel with @room", async () => {
```
Comment (Cadair) [inline src/tests/unit/substitutionsTest.ts]: Also `@here` and `@everyone`?

Comment (Half-Shot) [reply src/tests/unit/substitutionsTest.ts]: The title is wrong, should be replace @\room with @\channel. And Matrix only supports @\room

File: src/substitutions.ts (hunk 4 above)
Comment (Half-Shot) [reply src/substitutions.ts]: It is :(

**Comment by auscompgeek** [comment]
How will this behave if there are two hypothetical users with the display names `Alice` and `Alice Bob` in the same channel, for example?

**Comment by Half-Shot** [comment]
> How will this behave if there are two hypothetical users with the display names `Alice` and `Alice Bob` in the same channel, for example?

Good catch as always, I've provisioned for this now too.

**Pull request merged**